### ISSUE TITLE
Delete App Data

### ISSUE THREAD

**Issue by tflidd** [nextcloud/server#5539]
Use case:

- You tested an app and decided not to use it
- The app is not available any more for current NC versions

What I want:
- Delete all app-related data from user folders and database, e.g. an additional button on the web-interface where you enable and disable the app.

On top of that:
- for people with old setups, there are probably a lot of unused tables (I think some were cleaned during update procedures). Is there a way to find out which tables are used/not used in my current setup?

**Comment by e-alfred** [comment]
This is a very important feature because there are so many apps that come and go, sadly. A database/filesystem littered with old unused stuff that might create problems/conflicts with other apps/newer releases of Nextcloud is just no a great situation.

Aside from that, it would also get rid of mangling directly with the database if an app misbehaves and has to be restored to a "clean" state (i. e. removing everything related to the database/file system).

Anyway, before deleting something, the admin should be asked and a backup should be created in the file system somewhere of any deleted data (creating a dump of the table and a tarball with the files that were deleted). So, in case something important was stored in it, this wouldn't be lost. The only other option would be to restore the whole old installation from a backup and start it up somewhere (a virtual machine?) and grab the data. This would be a lot of work if only a certain table would be deleted for example.

**Comment by AND51** [comment]
Oh yes, this would be very nice. I use my Nextcloud since version 9 and I always updated it, tested this app and that app... I would like to clean up my Nextcloud, but I don't know how. Installing a fresh Nextcloud is no problem. Transferring files with rsync is also easy. But then I ran into problems, because I don't know which tables and files hold the users, the IDs, hashed passwords, contacts, calendar ... and so on. So a migration-tool or a cleanup-tool would be useful.

I tried using `php occ db:convert-type`. In one of the github issues here and in the documentation I read that this command will skip some unneeded tables. But when I execute this command, it says, "cannot migrate from mysql to mysql".

So this command is only for migration vom sqlite to mysql.

**Comment by schnappijedi** [comment]
Adding to this slightly (although is already known). The "oc_bruteforce_attempts" table appears to never be cleaned.

**Comment by MichaIng** [comment]
To keep this request alive, I would also love to see an ability to "purge" app data from database and data dir in case. Of course this requires a list of tables and files/dirs an app uses/creates.

**Comment by rakekniven** [comment]
I would for this as well.

During my support in Nextcloud forums I install many apps for testing and my db has a lot of garbage in it. Some tables I delete manually, with the risk to damage my system.

**Comment by benyanke** [comment]
This definitely needs to be done - it's subpar that an app installed and removed years ago still has table in my DB.

**Comment by rakekniven** [comment]
> This definitely needs to be done - it's subpar that an app installed and removed years ago still has table in my DB.

Like on every OS installation *SCNR*

**Comment by benyanke** [comment]
Yes, that’s a fault in legacy packaging methods too.

**Comment by phonon112358** [comment]
yes, to be able to purge any app would be a great feature...! 
some apps can be deleted via the web gui, but not all... this is very strange...

**Comment by rakekniven** [comment]
@phonon112358 Some apps are core apps and part of the installation. They cannot be uninstalled, just disabled.

**Comment by phonon112358** [comment]
@rakekniven yes, but I am talking about apps that I installed manually, e.g., default encryption module.

**Comment by brdobrdo** [comment]
I have an app that disabled web browser access to nextcloud. I can only use cpanel to access the nextcloud files so i Renamed the app folder in order to gain access from web browser again. Can't remove the app now...

**Comment by deezl** [comment]
OnlyOffice is completely broken now due to not being able to remove the app settings that do not get purged when deleting an app. This is really killing me. Please address this.

**Comment by ruedigerkupper** [comment]
Would like to have this, too.

**Comment by RuudschMaHinda** [comment]
Suggestion on how to implement this concerning UX (I also need this, badly!):

Have there be a checkmark that can be set when uninstalling an App. Maybe not in the list overview, but in the detail view of the app. Or after clicking uninstall, ask whether or not any data corresponding to this app should be removed from the database and/or filesystem. There might even be an option for a backup of that bit? Maybe perform a dump, or have an extra part of the database for that.

As for old apps that once had been installed, there could also be an option that maybe says that this app has previously been installed, and is using database resources, do you wish to clean up the database.

Or have it as an extra step after deactivating an app, before being able to click delete, it would prompt for database cleanup for all the tables that app (and only that app) created. Or the other way around, after having deleted the app, that option becomes available.

And of course, a general check would be nice too - that could basically check against all apps that ever were on the app-store, check which tables they (would) have created, and then offer deletion of said tables. All with double and triple checks if need be.

I started when it was still owncloud 5, switched as soon as there was the first stable fork - and it is such a hassle to keep the database clean.

**Comment by MichaIng** [comment]
Additionally `occ` CLI could get an `app:purge` command, similar to `apt-get remove` and `apt-get purge` to remove attached userdata and database entries.

I think Nextcloud cannot know which app creates which files and database tables, does it? In case I think it requires some new app API, e.g. an array that developers can use to add database tables and directories below userdata to be removed when using the purge option. If this array does not exist (yet), e.g. because the developer did not (yet) implement it or the app does not create anything such, the purge option could be hidden (to avoid confusion) or simply left without effect (means that admins might complain that the purge option left app-related stuff in place).

**Comment by IzzySoft** [comment]
> I think Nextcloud cannot know which app creates which files and database tables, does it?

AFAIK it does, at least for current apps which follow the guidelines. At least for database tables, apps needing their own are (were?) supposed to ship a `appinfo/database.xml` file defining its structures. Not sure if that's still the case, though: I've just checked on one of my installations, and found them missing from more recent apps I'm sure using their own tables.

**Comment by MichaIng** [comment]
Okay great, that makes it easier 👍.

**Comment by bes1002t** [comment]
I had to update nextcloud manually (replace all files) because of a bug I was not able to login to webgui anymore. But now some apps could be not updated together with nextcloud because there are no versions for 19.0.3 available e.g. files_rightclick. Now the installation says "The files of the app Right click (files_rightclick) were not replaced correctly. Make sure it is a version compatible with the server."

It would be cool to provide a method via occ to remove apps, so I don't have to use the webgui. Is this possible?

**Comment by MichaIng** [comment]
> It would be cool to provide a method via occ to remove apps, so I don't have to use the webgui. Is this possible?

You can already do that, check out the full list by calling `occ` without arguments and you'll find: `occ app:remove`

**Comment by bubelov** [comment]
> > I think Nextcloud cannot know which app creates which files and database tables, does it?
> 
> AFAIK it does, at least for current apps which follow the guidelines. At least for database tables, apps needing their own are (were?) supposed to ship a `appinfo/database.xml` file defining its structures. Not sure if that's still the case, though: I've just checked on one of my installations, and found them missing from more recent apps I'm sure using their own tables.

Same here. Just tried to disable -> remove the `news` app via the `occ` and, to my surprise, it wasn't a "clean" removal at all, leaving three tables in the DB (`oc_news_feeds`, `oc_news_folders`, `oc_news_items`). This isn't some kind of an obsolete app: it targets Nextcloud 20 and it also doesn't have `database.xml`.

I couldn't find a single place describing all of those tables. It seems like it's not that easy to connect apps and tables, in general. Makes me wonder what else the `remove` command might miss. Being able to do clean uninstalls is pretty important for certain use cases such as debugging and tweaking some apps. For instance, this `news` app crashed my Nextcloud 20 upgrade process yesterday and the logs say that something went wrong with the schema migration. This problem might not disappear if those tables and their schema are simply ignored by this "removal".

**Comment by RuudschMaHinda** [comment]
Here's an idea on how to implement NC "knowing" which app installed which files and tables (no idea how feasible it is):
Let NC save the current names of tables, and whenever an app is installed - check which new tables have been added. Save those new tables as linked with the app that just got installed. Therefor it should then be able to know which tables to offer to delete when uninstalling the app.

I hope that makes sense.
NC saves a list of the tables
App gets installed - NC checks and compares tables, stores the new tables as corresponding to app
Admin uninstalls app - NC asks whether or not to remove corresponding DB-entries, deletes tables according to what it saved as the diff when the app got installed -- there should of course be a warning, that user data linked with this app will be lost and gone if the apps should be reinstalled at a later date - hence it should ask about wanting to delete those entries.

That same list could be used for a cleanup of the DB. Which in turn can have checkmarks, and let the admin know which app-data is still lingering around. Maybe a checkbox for each app that created its own tables could be usefull here too.

Similarly possible with files - but that could take a bit more resources from the server, since a filesystem-check like that could take quite a bit longer with the amount of files and folders maybe increasing (think preview-app, or office-apps).

It could also make it possible to offer a "clean database" feature. If NC knows its own tables, creates a diff after an update for example, it can then offer to "reset" the database to only have the base entries be untouched. Similarly with old entries from ages ago. There could even be a resource somewhere on the store page, where the changes to the DB by different apps get reported by the NC-Apps-Installer to.

That way, the apps wouldn't even need to bring a database.xml file with them.

I just wish I could code properly to start trying out something like that.

**Comment by ruedigerkupper** [comment]
You'd need to do this when an app udaptes, too. (The update might introduce new tables.)

**Comment by MichaIng** [comment]
> NC checks and compares tables, stores the new tables as corresponding to app

This is kinda fragil. New tables could have been added for other reasons concurrently, as AFAIK Nextcloud does not go into strict maintenance mode during app installs/updates (?). Also one might want to include app-specific rows or columns, not only tables, e.g. for removing obsolete background jobs (probably handled differently already, but you get the point). And an app install could theoretically trigger other components to pre-create tables which are used not only by the installed app, like Deck creating a calendar and the calendar app/API then pre-creates a probably other missing table, e.g. if it was freshly installed as well.

Most robust would be if all apps must declare where they want to write any data to (file system or database) and if Nextcloud would not allow them to write anywhere else and on install/update checks for conflicts in case. Then on uninstall those declarations could be gone through to remove really everything where the app ever requested to write data to. But it becomes a bit complicated when thinking about internal API usage, e.g. special calendar entries, which are not done directly but through an API (I guess?), and if it is really wanted that those are removed or not. And in other cases apps might intentionally share tables or directories.

So most practically is likely an optional list for what is removed when "purging" the app, very similar to how APT and other package managers do it via post-removal scripts (`/var/lib/<name>/` content and such). It is then up to the app developers to implement/maintain those and up to the users to report when they find left-overs that are clearly only used by the particular app 😉.

**Comment by prilly-dev** [comment]
yes, when is this coming, my installation is filling up with crap that needs to be removed

**Comment by heeplr** [comment]
I miss some kind of "purge/clean DB" feature, too. Maybe considering the following might make sense in that concern:

Nextcloud is very easy and quick to setup. But migrating an old buggy/broken/bloated setup to a fresh install can be a pain. If that process would be as easy as the fresh setup itself, users could also use it to delete app data.

The process could look something like this:

1. Do a fresh install
2. export config and database dump from old setup
3. import config to the new setup + initialize database for that config (i.e. "mimic" manual configuration, app install etc.) ... (maybe even do some plausibility testing on the config.)
4. fill newly initialized & clean database with data from dump

That way the user should end up with a fresh install as if she had manually replicated all setup/configuration steps from the old install.

This would not only clean unused app data but also possible bugs/corruptions would be detected/fixed.

**Comment by MichaIng** [comment]
> 2. export config and database dump from old setup

And there is the tricky part: How can Nextcloud know which tables are required and which are leftovers from uninstalled apps. It's still an internal change required that strictly associates database tables with apps so that obsolete tables can be identified. Whether it is used then to do a "clean" export or to clean the database on a life instance as a repair step, or as part of a "delete data" button aside of the "disable"/"uninstall" buttons in the app store, like phpBB has it, is probably the smallest task.

Aside of this, `occ maintenance:repair` works fine to repair many usual issues with the database and `mysqldump` (when MySQL/MariaDB is used) to transfer the database (along with the data directory) form one system to another one to skip the NC-internal setup/configuration steps.

**Comment by heeplr** [comment]
> How can Nextcloud know which tables are required and which are leftovers from uninstalled apps

With the way I have in mind it wouldn't need to know. Just skip the stuff that is disabled in the output of  `occ config:list`.
When nextcloud imports data for all installed apps, it would automatically exclude disabled stuff (since it's not installed/initialized in the first place)

> Aside of this, `occ maintenance:repair` works fine to repair many usual issues

It doesn't solve all issues a fresh install solves and afais is meant to operate on a running instance. 

The "import config" and "import data" feature i'm talking about is more of a migration help. Removing cruft and fixing corruptions would just be a side effect. After all, starting from a fresh install is the last resort for fixing bugs/corruptions, even if that process would be super easy.

**Comment by MichaIng** [comment]
> Just skip the stuff that is disabled

What do you mean by "the stuff" and how does Nextcloud know which particular data base table are "the stuff" of which app? 😉
Same issue on import: Nextcloud does not know which table from the database dump belongs to which app, and the apps have not API that would tell Nextcloud. AFAIK, there is not even a central list of tables that belong to the Nextcloud core itself, which could be used for filtering.

Yes, current repair steps require a running instance and do only repair specific known issues and migration steps in particular.

I totally agree with you, that such a clean/repair/export/import feature would be awesome, but it requires a database table-to-app and file-to-app association, which currently does not exist. In a perfect world, all apps would require to define a list of database tables (or even single rows/entries, ah no, in a perfect world, this would be combined with the existing database API to avoid maintenance) and files/directories inside of the NC data directory, which they "own" (i.e. nothing else requires), so that on uninstall it is possible to parse this list and remove all contained database tables, files and directories safely, without breaking anything else. Collisions could be prevented in the first place by this as well, as a side effect. The other way round, an API could then exist to check where a particular file or table belongs to, and identify orphaned ones by this, to clean.

Don't understand me wrong, we're dreaming of the same thing, but looking into the way how apps define and handle their database entries and data files, it seems to be not trivial to implement, especially since one needs to decide whether to do it via internal database and file API changes, to have automatically/forcefully a list of all tables/files an app creates, or whether to do it as voluntary purge list that app developers may integrate, but without a guarantee that it is complete or even present.

**Comment by heeplr** [comment]
> What do you mean by "the stuff"

Apps & Features that leave cruft. (I'm not sure if that restricts to apps. I could imagine that e.g. theming or i18n could also leave leftovers when things are enabled/disabled)

> how does Nextcloud know which particular database table are "the stuff" of which app

When the config is imported, you check the database layout before and after you make a change. 

e.g. Config says that "passwords" was enabled. -> The app is downloaded, installed and initialized as if done manually. -> Then it should be trivial to check which tables were added. (I assume that apps create all tables upon install and not later on. In that case it wouldn't work)

> Tables and files are created via an API

It's perfectly possible that the API would need to be extended. But I'd be surprised if nextcloud couldn't access a list of all it's tables.

<details><summary>offtopic</summary>

> In a perfect world, all apps would require to define a list of database tables

Indeed. In a perfect world an app wouldn't be allowed to create or access any table if it's not mentioned in the manifest.
A decent registry to map the database landscape would contribute to general structure, security and maintenance.
</details>

**Comment by MichaIng** [comment]
> When the config is imported, you check the database layout before and after you make a change.

Well, the config does not import app configurations and data, nor do apps create all their tables and files/dirs directly when being installed or enabled or even used. So you'll likely end up with some configs and data missing, not only obsolete/orphaned ones. Moreover, even when freshly installing an app and comparing tables prior to and after install, there are lots of other reasons why a table might have changed or been added, including that the app triggered a core feature, so that the added table belongs not to the app but the core and may be used later by other apps as well. So the comparison approach sounds highly fragile and dangerous to me 😢.

> I assume that apps create all tables upon install and not later on.

I'm pretty sure that's not the case, but I haven't verified it.

> It's perfectly possible that the API would need to be extended.

This is somehow the core of the issue: Of course there is an API to define a table namespace and create/write to tables etc, but an app does not need to initialise all tables on install or when being enabled but can do that on demand, calling the API at any function when using any of it's future of settings section.

> In a perfect world an app wouldn't be allowed to create or access any table if it's not mentioned in the manifest.

👍 indeed. That manifest could be combined with what apps currently do like here: https://github.com/nextcloud/deck/blob/master/lib/Db/Attachment.php (not standardised)
If such database namespace, key and type definitions would need to be done in a standardised way, so that the core would be able to parse those to get all tables "owned" by that app, that would be great.

But probably there is already more possible than I can see. I'm not good in understanding the Nextcloud code I have to admit, I just compare the name spaces, function names and where they are called, and that does not look like it's standardised enough that the core could get the info it needs in a generic way.

Isn't there any app developer around here that could shed some light on this topic, so that we know better where we currently stand? 😄

**Comment by heldzz** [comment]
My current case lead to this topic: Some notifications cannot be dismissed (mainly suspicious logins; "Failed to dismiss notification"), so I just wanted to remove and re-install that app. Since it is a "shipped" app, it can't be removed, so I searched for a way to clear those sticking database entries. I don't want to manually mess around in the database, which once was the reason that I had to setup the whole nextcloud from scrath...
Any updated about this topic?

**Comment by prilly-dev** [comment]
Sorry it seems that the database cleaning is still missing, my database is filled with crap that cant be removed without manualy messing with it, it think i have around 20% junk of the total used space
-------- Opprinnelig melding --------Fra: heldzz ***@***.***> Dato: 20.02.2022  18:05  (GMT+01:00) Til: nextcloud/server ***@***.***> Ko: prilly-dev ***@***.***>, Comment ***@***.***> Emne: Re: [nextcloud/server] Delete App Data (#5539) 
My current case lead to this topic: Some notifications cannot be dismissed (mainly suspicious logins; "Failed to dismiss notification"), so I just wanted to remove and re-install that app. Since it is a "shipped" app, it can't be removed, so I searched for a way to clear those sticking database entries. I don't want to manually mess around in the database, which once was the reason that I had to setup the whole nextcloud from scrath...
Any updated about this topic?

**Comment by kilrau** [comment]
I am also missing this purge feature, not only for cleaning up my quite messy years-old db, but also to get apps with faulty settings back to work, right now one has to `config:list appname` & then manually delete all settings of an app manually via `config:app:delete appname settingname` as a workaround :expressionless:

**Comment by janaxhell** [comment]
> I am also missing this purge feature, not only for cleaning up my quite messy years-old db, but also to get apps with faulty settings back to work, right now one has to `config:list appname` & then manually delete all settings of an app manually via `config:app:delete appname settingname` as a workaround 😑

May I ask how can I do this in Nextcloud AIO for GpodderSync? All I get with

`docker exec --user www-data -it nextcloud-aio-nextcloud php occ config:list gpoddersync `

is
```
{
    "apps": {
        "gpoddersync": {
            "installed_version": "3.8.0",
            "gpoddersync": "",
            "enabled": "yes",
            "types": ""
        }
    }
}
```
So I have no idea what to delete, the command doesn't show a list, but what looks like the content of a config file.

**Comment by HyperCriSiS** [comment]
I have an issue with an app and was searching for a way to delete its data. I thought this is such a trivial thing, this must be somewhere and I must be blind not to find it. :neutral_face:

**Comment by RuudschMaHinda** [comment]
So far:
backup your database, then delete all the tables in it that bare the name of the app after having deactivated and deleted the app.

**Comment by D-side** [comment]
I've spent some time poking around the filesystem and database and I've settled on this list:

- Database tables mentioned in the app's `lib/Migrations`
- Entries in the configuration table (don't recall which one, but it contains settings from all apps)
- Leftover files in the file system (users can remove files in their spaces themselves, but some may be outside)
- (Once) A change to the list of MIME types

---

> delete all the tables in it that bare the name of the app

Sadly, I don't believe Nextcloud enforces this in any way, a more reliable approach would be to open up the respective app's source code and look through files in `lib/Migrations` to see what tables they create. E. g. I spent a while looking which app does the `favorites` table belong to — much to my surprise, it was [radio](https://git.project-insanity.org/onny/nextcloud-app-radio/-/blob/master/lib/Migration/Version000000Date20181013124731.php).  (Also, *bear)

_(I was hoping there'd be a consolidated schema file a la `db/schema.rb` in Ruby on Rails, but no such luck, though I suppose separating the app's facilities from base Nextcloud's could be difficult to implement.)_

**Comment by dartcafe** [comment]
I stumbled over this issue comming from #28283. 

In Polls we implemented `occ polls:db:purge` for removing
* all migration records, 
* all app configs and
* all tables (oc_polls_*) 

regarding Polls. But the app must be installed and activated for that.

It is not ideal, but it works for getting rid of all Polls related db entries. 

Proposal: 
* Apps have to provide a method for removing the app's tables.
* Nextcloud takes care of cleaning other Nextcloud tables from the app's data (i.e. `migrations`, `app_config`, etc.)

This could be triggered with an extra command i.e. `occ app:purge` in contrast to `occ app:remove`, which only removes the app's sources, while `purge` also removes the app's tables.

This way app developers have to take care of removing their tables and nextcloud itself cares for all other records in the core tables.

**Comment by sepo27** [comment]
Bump

**Comment by PilaScat** [comment]
bump

**Comment by infinitytec** [comment]
Same issue here. I put a bad value in the settings for Assistant and the web interface breaks so I can't reset it. For now I am just disabling the app with OCC.

**Comment by mwbreur** [comment]
Late to the party but i did the following:
Issue with onlyoffice : 
(Through occweb)
app:list ->onlyoffice is not listed
config:list 
-> apps - onlyoffice - occ $  config:list onlyoffice
{    
"apps": {       
 "onlyoffice": {
   "DocumentServerUrl": "https:\/\/nc.domain.eu\/index.php\/apps\/documentserver_community\/",            
   "cronChecker": "true",           
   "jwt_header": "AuthorizationJwt"       
  }   
 }
}

settings still there.

**Comment by someone-somenet-org** [comment]
Bump.

(Nextcloud needs so much more automatic and manual housekeeping features)

**Comment by EdithBirrer1** [comment]
Agreed - I had an issue with OnlyOffice, and removing the app does not help, because the settings are persistent. 
A command to remove all settings for an app would be very helpful! The above suggestion of app:purge sounds reasonable, I support it.

**Comment by TimvanDijkNL** [comment]
This issue has been open for.. 8 years now ? 🤨

I am having the same issue...

**Comment by miro-zamiro** [comment]
Please add it!...

**Comment by adva-webmaster** [comment]
I still miss the app purge in Version 31. Will update to 33, but there is only few hope for improvement...
